graderutils_format/templates/feedback.html

```diff
@@ -42,12 +42,16 @@ <h3 class="testgroup-header">{{ result_group.title }}</h3>
         <div class="panel-heading">
             <h5>
                 <a data-toggle="collapse"
-                   href="#result{{ result_group_loop.index }}-{{ test_result_loop.index }}">{{ result.title }}</a>
+                    href="#result{{ result_group_loop.index }}-{{ test_result_loop.index }}">{{ result.title }}</a>
+                <span> </span>
                 {% if result.status == "passed" %}
+                <span class="badge badge-success">{{ result.points }} / {{ result.maxPoints }}</span>
                 <span class="label feedback-label label-success">Success</span>
                 {% elif result.status == "failed" %}
+                <span class="badge badge-danger">{{ result.points }} / {{ result.maxPoints }}</span>
                 <span class="label feedback-label label-danger">Failed</span>
                 {% elif result.status == "error" %}
+                <span class="badge badge-dark">{{ result.points }} / {{ result.maxPoints }}</span>
```

Comment:
Try to use default bootstrap labels. I would avoid adding much custom css stuff as all of it is inlined to the feedback, which is stored for every submission.

A+ api basically allows using bootstrap 3, so you beed to change these to "label": https://getbootstrap.com/docs/3.3/components/#labels

I think label-success, label-danger, and label-default should be ok for these.

For reference, issue about upgrading the CSS for A+ protocol (or namespacing them) https://github.com/apluslms/a-plus/issues/584

Reply:
I just found the conflicting part in the bootstrap stylesheet:
`.panel-default>.panel-heading .badge{color:#f5f5f5;background-color:#333}`
But I don't know how to fix this overriding default `badge`.

Also the normal labels such as `label-success` etc. have less rounded corners, which doesn't fit the style that A+ has where points are always shown as having rounded corners around them...

Reply:
> I just found the conflicting part in the bootstrap stylesheet:
> `.panel-default>.panel-heading .badge{color:#f5f5f5;background-color:#333}`
> But I don't know how to fix this overriding the default `badge`.
> 
> Also the normal labels such as `label-success` etc. have less rounded corners, which doesn't fit the style that A+ has where points are always shown as having rounded corners around them...

I found a fix for the default badge color inside panel-heading! Can we use that instead of the ugly label-success etc., which don't have rounded corners?

Reply:
The targeted CSS rule in the panel heading (`div.panel-default>.panel-heading .badge `) sounds good to me. It is also nice that it shouldn't affect anything in unwanted places since it is so strictly targeted to this case. As a sidenote, I think that the usual coding style is to have spaces around the `>` operator.

Reply:
> The targeted CSS rule in the panel heading (`div.panel-default>.panel-heading .badge `) sounds good to me. It is also nice that it shouldn't affect anything in unwanted places since it is so strictly targeted to this case. As a sidenote, I think that the usual coding style is to have spaces around the `>` operator.

Fixed, thanks!

Reply:
`div.panel-default > .panel-heading` targets all bootstrap panels, so it's pretty wide. `.exercise-content > .panel-heading > ` should be more ok. Interestingly, we don't actually have good anchors for feedback styles. I guess every feedback could generate a random prefix and use that (e.g. top most element in the feedback would have a class like `python-grader-utils-v3-<8 random characters>`. All styles would be defined to be sub elements of that.

However, I not so fond of including styles per feedback. Specially things that change default bootstrap styles for the hole page.. I think all css rules added per feedback should be checked and reconsidered if they can be moved elsewhere (A+ basically) or just dropped.

That said, `badge-*` classes are custom additions in A+ and not part of bootstrap 3. We have two okish solution:
    
* Specify that `badge-*` classes are part of A+ protocol and include the missing classes in A+ css.
* Specify that only bootstrap 3 styles are part of A+ protocol with addition of `label-pill`, which would mimic the behaviour of `badge-pill` from bootstrap 4 (which could be used in A+ too to make badge like labels).

For reference:

* css for `.label-pill`

```css
.label-pill {
  padding-right: .6em;
  padding-left: .6em;
  border-radius: 10rem;
}
```

* example of really bad css in grader utils :( :
    
```csss
pre {
    white-space: pre-wrap;
}
```

Reply:
Major changes to the CSS are not in the scope of this PR. I think it is OK to fix the points badge style issue like Jimmy has done since it does not add new CSS classes. Fine-tuning the CSS selector is of course a good idea, so it is worth trying out the suggested `.exercise-content`. Otherwise, let's stay on topic in this PR and not revamp all CSS at once.

Reply:
That said, I don't consider `badge-*` to be part of the A+ protocol. Bootstrap 3 quite clearly should be assumed to be. Thus, it's possible that these classes will stop working at some point. It's hard to ensure that those classes are not refactored away from A+. However, it's unlikely that it would happen before there is good alternative here (e.g. prefixed bootstrap 4 classes).

So, I would be fine as long as the custom css doesn't touch normal badge classes currently used in A+ (i.e. appropriate anchor is used).

Reply:
Old feedback in old courses will definitely break at some point since feedback is arbitrary HTML. That's OK. People don't often open feedback in old courses and we have to make reasonable compromises in order to improve the system for the future.

Reply:
> Major changes to the CSS are not in the scope of this PR. I think it is OK to fix the points badge style issue like Jimmy has done since it does not add new CSS classes. Fine-tuning the CSS selector is of course a good idea, so it is worth trying out the suggested `.exercise-content`. Otherwise, let's stay on topic in this PR and not revamp all CSS at once.

I tried `.exercise-content > .panel-heading .badge` but that didn't work. Instead I got both `.exercise-content .panel-heading .badge` and `div.grading-task .panel-default > .panel-heading .badge` to work.

Which one should we use?

Reply:
`.exercise-content .panel-heading .badge` looks right. I made a mistake when I read the dom when crafting the example (i.e. `.panel-heading` is not direct child of `.exercise-content`)

Reply:
`div.grading-task` is probably only used by python-grader-utils, so including that in the CSS rule would narrow the scope to python-grader-utils.

I have a feeling that `.exercise-content` is only included in the chapter exercise pages, not in traditional independent exercise pages. Does it work in the teacher's "inspect submission" page?

Reply:
> `div.grading-task` is probably only used by python-grader-utils, so including that in the CSS rule would narrow the scope to python-grader-utils.
> 
> I have a feeling that `.exercise-content` is only included in the chapter exercise pages, not in traditional independent exercise pages. Does it work in the teacher's "inspect submission" page?

You're correct. `.exercise-content .panel-heading .badge` did not work in teacher's inspect submission page. However, `div.grading-task .panel-default > .panel-heading .badge` did. 
So, I guess we're using that one then?

Reply:
OK for me.

Reply:
I'm not happy with it, but I guess it's best we can do from this components point of view. I hope A+/MOOC-Grader parts get better so we need less hackish stuff here.